File: balanced-brackets.js
var balanced_brackets = function(str){
	var length = str.length;

	if(length < 2 || length%2 !== 0){
		console.log('NO');
	}
	else {

		var open_brackets = new Array();
		var brackets_map = {'[':']', '(':')', '{':'}'};
		var open_brackets_set = {'[':1,'{':1,'(':1};
		var last_open = '';
		var expected_bracket = '';
		var is_balanced = true;

		for(let i = 0; i < length; i++){
			if(open_brackets_set.hasOwnProperty(str[i])){
				open_brackets.push(str[i]);
			}
			else {
				last_open = open_brackets.pop();
				expected_bracket = brackets_map[last_open];

				if(expected_bracket !== str[i]){
					is_balanced = false;
					break;
				}
			}
		}

		(is_balanced && open_brackets.length === 0) ? console.log('YES') : console.log('NO');
	}
}

balanced_brackets('');

File: balanced-brackets.py
#!/bin/python3

import sys

def isBalanced(s):
    # Complete this function
    str_len = len(s)
    if((str_len < 2) or (str_len % 2 != 0)):
        return "NO"
    
    brackets_map = {')':'(', ']':'[', '}':'{'}
    open_brckts = {'{','(','['}
    stack = list()
    bracket = ""
    is_balanced = True
    for i in range(str_len):
        bracket = s[i]
        if(bracket in open_brckts):
            stack.append(bracket)
        else:
            last_open = brackets_map[bracket]
            if(not stack):
               return "NO" 
            if(last_open != stack.pop()):
                is_balanced = False
                break
               
    
    if(is_balanced and not stack):
        return "YES"
    else:
        return "NO"

if __name__ == "__main__":
    t = int(input().strip())
    for a0 in range(t):
        s = input().strip()
        result = isBalanced(s)
        print(result)
